src_feedback/elf.rs

```diff
@@ -0,0 +1,666 @@
+
+	let me = self.backobj.borrow();
+
+	Ok(me.ehdr.as_ref().unwrap().e_type)
+    }
+
+    fn check_section_index(&self, sect_idx: usize) -> Result<(), Error> {
+	let nsects = self.get_num_sections()?;
+
+	if nsects <= sect_idx {
+	    return Err(Error::new(ErrorKind::InvalidInput, "the index is too big"));
+	}
+	Ok(())
+    }
+
+    pub fn section_seek(&self, sect_idx: usize) -> Result<(), Error> {
+	self.check_section_index(sect_idx)?;
+	self.ensure_shdrs()?;
+	let me = self.backobj.borrow();
+	read_elf_section_seek(&mut *self.file.borrow_mut(), &me.shdrs.as_ref().unwrap()[sect_idx])
+    }
+
+    pub fn section_offset_seek(&self, sect_idx: usize, offset: usize) -> Result<(), Error> {
+	self.check_section_index(sect_idx)?;
+	self.ensure_shdrs()?;
+	let me = self.backobj.borrow();
+	read_elf_section_offset_seek(&mut *self.file.borrow_mut(), &me.shdrs.as_ref().unwrap()[sect_idx], offset)
+    }
+
+    /// Read the raw data of the section of a given index.
+    pub fn read_section_raw(&self, sect_idx: usize) -> Result<Vec<u8>, Error> {
+	self.check_section_index(sect_idx)?;
+	self.ensure_shdrs()?;
+
+	let me = self.backobj.borrow();
+	read_elf_section_raw(&mut *self.file.borrow_mut(), &me.shdrs.as_ref().unwrap()[sect_idx])
+    }
+
+    /// Get the name of the section of a given index.
+    pub fn get_section_name(&self, sect_idx: usize) -> Result<&str, Error> {
+	self.check_section_index(sect_idx)?;
+
+	self.ensure_shstrtab()?;
+
+	let me = self.backobj.borrow();
+
+	let sect = &me.shdrs.as_ref().unwrap()[sect_idx];
+	let name = get_elf_section_name(sect, unsafe { (*self.backobj.as_ptr()).shstrtab.as_ref().unwrap() });
+	if name.is_none() {
+	    return Err(Error::new(ErrorKind::InvalidData, "invalid section name"));
+	}
+	Ok(name.unwrap())
+    }
+
+    pub fn get_section_size(&self, sect_idx: usize) -> Result<usize, Error> {
+	self.check_section_index(sect_idx)?;
+	self.ensure_shdrs()?;
+
+	let me = self.backobj.borrow();
+	let sect = &me.shdrs.as_ref().unwrap()[sect_idx];
+	Ok(sect.sh_size as usize)
+    }
+
+    pub fn get_num_sections(&self) -> Result<usize, Error> {
+	self.ensure_ehdr()?;
+	let me = self.backobj.borrow();
+	Ok(me.ehdr.as_ref().unwrap().e_shnum as usize)
+    }
+
+    /// Find the section of a given name.
+    ///
+    /// This function return the index of the section if found.
+    pub fn find_section(&self, name: &str) -> Result<usize, Error> {
+	let nsects = self.get_num_sections()?;
+	for i in 0..nsects {
+	    if self.get_section_name(i).unwrap() == name {
+		return Ok(i);
+	    }
+	}
+	Err(Error::new(ErrorKind::NotFound, format!("Does not found the give section: {}", name)))
+    }
+
+    pub fn find_symbol(&self, address: u64, st_type: u8) -> Result<(&str, u64), Error> {
+	self.ensure_symtab()?;
+	self.ensure_strtab()?;
+
+	let me = self.backobj.borrow();
+	let idx_r = search_address_opt_key(me.symtab.as_ref().unwrap(), address, &|sym: &Elf64_Sym| {
+	    if sym.st_info & 0xf != st_type || sym.st_shndx == SHN_UNDEF {
```

Comment:
Does sym.get_type() do the same thing as "sym.st_info & 0xf"?

Reply:
Yes, we should use get_type() instead.